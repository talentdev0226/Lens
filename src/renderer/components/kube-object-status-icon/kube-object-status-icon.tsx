/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import "./kube-object-status-icon.scss";

import React from "react";
import { Icon } from "../icon";
import { cssNames, formatDuration } from "../../utils";
import { KubeObject, KubeObjectStatus, KubeObjectStatusLevel } from "../../..//extensions/renderer-api/k8s-api";
import { KubeObjectStatusRegistry } from "../../../extensions/registries";

function statusClassName(level: KubeObjectStatusLevel): string {
  switch (level) {
    case KubeObjectStatusLevel.INFO:
      return "info";
    case KubeObjectStatusLevel.WARNING:
      return "warning";
    case KubeObjectStatusLevel.CRITICAL:
      return "error";
  }
}

function statusTitle(level: KubeObjectStatusLevel): string {
  switch (level) {
    case KubeObjectStatusLevel.INFO:
      return "Info";
    case KubeObjectStatusLevel.WARNING:
      return "Warning";
    case KubeObjectStatusLevel.CRITICAL:
      return "Critical";
  }
}

function getAge(timestamp: string) {
  return timestamp
    ? formatDuration(Date.now() - new Date(timestamp).getTime(), true)
    : "";
}

interface SplitStatusesByLevel {
  maxLevel: string;
  criticals: KubeObjectStatus[];
  warnings: KubeObjectStatus[];
  infos: KubeObjectStatus[];
}

/**
 * This function returns the class level for corresponding to the highest status level
 * and the statuses split by their levels.
 * @param src a list of status items
 */
function splitByLevel(src: KubeObjectStatus[]): SplitStatusesByLevel {
  const parts = new Map(Object.values(KubeObjectStatusLevel).map(v => [v, []]));

  src.forEach(status => parts.get(status.level).push(status));

  const criticals = parts.get(KubeObjectStatusLevel.CRITICAL);
  const warnings = parts.get(KubeObjectStatusLevel.WARNING);
  const infos = parts.get(KubeObjectStatusLevel.INFO);
  const maxLevel = statusClassName(criticals[0]?.level ?? warnings[0]?.level ?? infos[0].level);

  return { maxLevel, criticals, warnings, infos };
}

interface Props {
  object: KubeObject;
}

export class KubeObjectStatusIcon extends React.Component<Props> {
  renderStatuses(statuses: KubeObjectStatus[], level: number) {
    const filteredStatuses = statuses.filter((item) => item.level == level);

    return filteredStatuses.length > 0 && (
      <div className={cssNames("level", statusClassName(level))}>
        <span className="title">
          {statusTitle(level)}
        </span>
        {
          filteredStatuses.map((status, index) => (
            <div key={`kube-resource-status-${level}-${index}`} className={cssNames("status", "msg")}>
              - {status.text} <span className="age"> Â· {getAge(status.timestamp)}</span>
            </div>
          ))
        }
      </div>
    );
  }

  render() {
    const statuses = KubeObjectStatusRegistry.getInstance().getItemsForObject(this.props.object);

    if (statuses.length === 0) {
      return null;
    }

    const { maxLevel, criticals, warnings, infos } = splitByLevel(statuses);

    return (
      <Icon
        material={maxLevel}
        className={cssNames("KubeObjectStatusIcon", maxLevel)}
        tooltip={{
          children: (
            <div className="KubeObjectStatusTooltip">
              {this.renderStatuses(criticals, KubeObjectStatusLevel.CRITICAL)}
              {this.renderStatuses(warnings, KubeObjectStatusLevel.WARNING)}
              {this.renderStatuses(infos, KubeObjectStatusLevel.INFO)}
            </div>
          ),
        }}
      />
    );
  }
}
