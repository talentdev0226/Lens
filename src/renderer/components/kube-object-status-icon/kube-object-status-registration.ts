/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { KubeObject } from "../../../common/k8s-api/kube-object";
import type { KubeObjectStatus } from "../../../extensions/renderer-api/kube-object-status";

export interface KubeObjectStatusRegistration {
  kind: string;
  apiVersions: string[];
  resolve: (object: KubeObject) => KubeObjectStatus;
}
