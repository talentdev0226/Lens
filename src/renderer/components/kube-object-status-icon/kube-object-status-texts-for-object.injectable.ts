/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { getInjectable, lifecycleEnum } from "@ogre-tools/injectable";
import kubeObjectStatusTextsInjectable from "./kube-object-status-texts.injectable";
import type { KubeObject } from "../../../common/k8s-api/kube-object";
import { conforms, eq, includes } from "lodash/fp";
import type { KubeObjectStatusRegistration } from "./kube-object-status-registration";
import { computed } from "mobx";

const kubeObjectStatusTextsForObjectInjectable = getInjectable({
  id: "kube-object-status-texts-for-object",

  instantiate: (di, kubeObject: KubeObject) => {
    const allStatusTexts = di.inject(kubeObjectStatusTextsInjectable);

    return computed(() =>
      allStatusTexts
        .get()
        .filter(toKubeObjectRelated(kubeObject))
        .map(toStatus(kubeObject))
        .filter(Boolean),
    );
  },

  lifecycle: lifecycleEnum.keyedSingleton({
    getInstanceKey: (di, kubeObject: KubeObject) => kubeObject.getId(),
  }),
});

const toKubeObjectRelated = (kubeObject: KubeObject) =>
  conforms({
    kind: eq(kubeObject.kind),
    apiVersions: includes(kubeObject.apiVersion),
  });

const toStatus =
  (kubeObject: KubeObject) => (item: KubeObjectStatusRegistration) =>
    item.resolve(kubeObject);

export default kubeObjectStatusTextsForObjectInjectable;
