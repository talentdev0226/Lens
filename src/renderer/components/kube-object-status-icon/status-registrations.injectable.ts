/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { getInjectable, lifecycleEnum } from "@ogre-tools/injectable";
import { computed } from "mobx";
import rendererExtensionsInjectable from "../../../extensions/renderer-extensions.injectable";

const statusRegistrationsInjectable = getInjectable({
  id: "status-registrations",

  instantiate: (di) => {
    const extensions = di.inject(rendererExtensionsInjectable);

    return computed(() =>
      extensions.get().flatMap((extension) => extension.kubeObjectStatusTexts),
    );
  },

  lifecycle: lifecycleEnum.singleton,
});

export default statusRegistrationsInjectable;
