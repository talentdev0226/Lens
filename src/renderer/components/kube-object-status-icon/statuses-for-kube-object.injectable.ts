/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { getInjectable, lifecycleEnum } from "@ogre-tools/injectable";
import statusRegistrationsInjectable from "./status-registrations.injectable";
import type { KubeObject } from "../../../common/k8s-api/kube-object";
import { conforms, eq, includes } from "lodash/fp";
import type { KubeObjectStatusRegistration } from "./kube-object-status-registration";

const statusesForKubeObjectInjectable = getInjectable({
  id: "statuses-for-kube-object",

  instantiate: (di, kubeObject: KubeObject) =>
    di
      .inject(statusRegistrationsInjectable)
      .get()
      .filter(toKubeObjectRelated(kubeObject))
      .map(toStatus(kubeObject))
      .filter(Boolean),

  lifecycle: lifecycleEnum.transient,
});

const toKubeObjectRelated = (kubeObject: KubeObject) =>
  conforms({
    kind: eq(kubeObject.kind),
    apiVersions: includes(kubeObject.apiVersion),
  });

const toStatus =
  (kubeObject: KubeObject) => (item: KubeObjectStatusRegistration) =>
    item.resolve(kubeObject);

export default statusesForKubeObjectInjectable;
