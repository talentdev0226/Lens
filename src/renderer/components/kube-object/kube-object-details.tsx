import "./kube-object-details.scss";

import React from "react";
import { disposeOnUnmount, observer } from "mobx-react";
import { computed, observable, reaction } from "mobx";
import { createPageParam, navigation } from "../../navigation";
import { Drawer } from "../drawer";
import { KubeObject } from "../../api/kube-object";
import { Spinner } from "../spinner";
import { apiManager } from "../../api/api-manager";
import { crdStore } from "../+custom-resources/crd.store";
import { CrdResourceDetails } from "../+custom-resources";
import { KubeObjectMenu } from "./kube-object-menu";
import { kubeObjectDetailRegistry } from "../../api/kube-object-detail-registry";

/**
 * Used to store `object.selfLink` to show more info about resource in the details panel.
 */
export const kubeDetailsUrlParam = createPageParam({
  name: "kube-details",
  isSystem: true,
});

/**
 * Used to highlight last active/selected table row with the resource.
 *
 * @example
 * If we go to "Nodes (page) -> Node (details) -> Pod (details)",
 * last clicked Node should be "active" while Pod details are shown).
 */
export const kubeSelectedUrlParam = createPageParam({
  name: "kube-selected",
  isSystem: true,
  get defaultValue() {
    return kubeDetailsUrlParam.get();
  }
});

export function showDetails(selfLink = "", resetSelected = true) {
  const detailsUrl = getDetailsUrl(selfLink, resetSelected);

  navigation.merge({ search: detailsUrl });
}

export function hideDetails() {
  showDetails();
}

export function getDetailsUrl(selfLink: string, resetSelected = false, mergeGlobals = true) {
  const params = new URLSearchParams(mergeGlobals ? navigation.searchParams : "");

  params.set(kubeDetailsUrlParam.urlName, selfLink);

  if (resetSelected) {
    params.delete(kubeSelectedUrlParam.urlName);
  } else {
    params.set(kubeSelectedUrlParam.urlName, kubeSelectedUrlParam.get());
  }

  return `?${params}`;
}

export interface KubeObjectDetailsProps<T = KubeObject> {
  className?: string;
  object: T;
}

@observer
export class KubeObjectDetails extends React.Component {
  @observable isLoading = false;
  @observable.ref loadingError: React.ReactNode;

  @computed get path() {
    return kubeDetailsUrlParam.get();
  }

  @computed get object() {
    const store = apiManager.getStore(this.path);

    if (store) {
      return store.getByPath(this.path);
    }
  }

  @computed get isCrdInstance() {
    return !!crdStore.getByObject(this.object);
  }

  @disposeOnUnmount
  loader = reaction(() => [
    this.path,
    this.object, // resource might be updated via watch-event or from already opened details
    crdStore.items.length, // crd stores initialized after loading
  ], async () => {
    this.loadingError = "";
    const { path, object } = this;

    if (!object) {
      const store = apiManager.getStore(path);

      if (store) {
        this.isLoading = true;

        try {
          await store.loadFromPath(path);
        } catch (err) {
          this.loadingError = <>Resource loading has failed: <b>{err.toString()}</b></>;
        } finally {
          this.isLoading = false;
        }
      }
    }
  });

  render() {
    const { object, isLoading, loadingError, isCrdInstance } = this;
    const isOpen = !!(object || isLoading || loadingError);
    let title = "";
    let details: React.ReactNode[];

    if (object) {
      const { kind, getName } = object;

      title = `${kind}: ${getName()}`;
      details = kubeObjectDetailRegistry.getItemsForKind(object.kind, object.apiVersion).map((item, index) => {
        return <item.components.Details object={object} key={`object-details-${index}`}/>;
      });

      if (isCrdInstance && details.length === 0) {
        details.push(<CrdResourceDetails object={object}/>);
      }
    }

    return (
      <Drawer
        className="KubeObjectDetails flex column"
        open={isOpen}
        title={title}
        toolbar={<KubeObjectMenu object={object} toolbar={true}/>}
        onClose={hideDetails}
      >
        {isLoading && <Spinner center/>}
        {loadingError && <div className="box center">{loadingError}</div>}
        {details}
      </Drawer>
    );
  }
}
