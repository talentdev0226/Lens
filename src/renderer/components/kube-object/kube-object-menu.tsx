import React from "react";
import { autobind, cssNames } from "../../utils";
import { KubeObject } from "../../api/kube-object";
import { editResourceTab } from "../dock/edit-resource.store";
import { MenuActions, MenuActionsProps } from "../menu/menu-actions";
import { hideDetails } from "./kube-object-details";
import { apiManager } from "../../api/api-manager";
import { kubeObjectMenuRegistry } from "../../../extensions/registries/kube-object-menu-registry";

export interface KubeObjectMenuProps<T> extends MenuActionsProps {
  object: T | null | undefined;
  editable?: boolean;
  removable?: boolean;
}

export class KubeObjectMenu<T extends KubeObject> extends React.Component<KubeObjectMenuProps<T>> {
  get store() {
    const { object } = this.props;

    if (!object) return;

    return apiManager.getStore(object.selfLink);
  }

  get isEditable() {
    const { editable } = this.props;

    return editable !== undefined ? editable : !!(this.store && this.store.update);
  }

  get isRemovable() {
    const { removable } = this.props;

    return removable !== undefined ? removable : !!(this.store && this.store.remove);
  }

  @autobind()
  async update() {
    hideDetails();
    editResourceTab(this.props.object);
  }

  @autobind()
  async remove() {
    hideDetails();
    const { object, removeAction } = this.props;

    if (removeAction) await removeAction();
    else await this.store.remove(object);
  }

  @autobind()
  renderRemoveMessage() {
    const { object } = this.props;

    if (!object) {
      return null;
    }

    return (
      <p>Remove {object.kind} <b>{object.getName()}</b>?</p>
    );
  }

  getMenuItems(object: T): React.ReactChild[] {
    if (!object) {
      return [];
    }

    return kubeObjectMenuRegistry
      .getItemsForKind(object.kind, object.apiVersion)
      .map(({components: { MenuItem }}, index) => (
        <MenuItem
          object={object}
          key={`menu-item-${index}`}
          toolbar={toolbar}
        />
      ));
  }

  render() {
    const { remove, update, renderRemoveMessage, isEditable, isRemovable } = this;
    const { className, object, editable, removable, ...menuProps } = this.props;

    return (
      <MenuActions
        className={cssNames("KubeObjectMenu", className)}
        updateAction={isEditable ? update : undefined}
        removeAction={isRemovable ? remove : undefined}
        removeConfirmationMessage={renderRemoveMessage}
        {...menuProps}
      >
        {this.getMenuItems(object)}
      </MenuActions>
    );
  }
}
