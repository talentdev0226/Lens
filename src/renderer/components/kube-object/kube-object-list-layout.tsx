import React from "react";
import { computed } from "mobx";
import { disposeOnUnmount, observer } from "mobx-react";
import { cssNames } from "../../utils";
import { KubeObject } from "../../api/kube-object";
import { ItemListLayout, ItemListLayoutProps } from "../item-object-list/item-list-layout";
import { KubeObjectStore } from "../../kube-object.store";
import { KubeObjectMenu } from "./kube-object-menu";
import { kubeSelectedUrlParam, showDetails } from "./kube-object-details";
import { kubeWatchApi } from "../../api/kube-watch-api";
import { clusterContext } from "../context";

export interface KubeObjectListLayoutProps extends ItemListLayoutProps {
  store: KubeObjectStore;
  dependentStores?: KubeObjectStore[];
}

@observer
export class KubeObjectListLayout extends React.Component<KubeObjectListLayoutProps> {
  @computed get selectedItem() {
    return this.props.store.getByPath(kubeSelectedUrlParam.get());
  }

  componentDidMount() {
    const { store, dependentStores = [] } = this.props;
    const stores = Array.from(new Set([store, ...dependentStores]));

    disposeOnUnmount(this, [
      kubeWatchApi.subscribeStores(stores, {
        preload: true,
        namespaces: clusterContext.contextNamespaces,
      })
    ]);
  }

  onDetails = (item: KubeObject) => {
    if (this.props.onDetails) {
      this.props.onDetails(item);
    } else {
      showDetails(item.selfLink);
    }
  };

  render() {
    const { className, store, items = store.contextItems, ...layoutProps } = this.props;

    return (
      <ItemListLayout
        {...layoutProps}
        className={cssNames("KubeObjectListLayout", className)}
        store={store}
        items={items}
        preloadStores={false} // loading handled in kubeWatchApi.subscribeStores()
        detailsItem={this.selectedItem}
        onDetails={this.onDetails}
        renderItemMenu={(item) => {
          return <KubeObjectMenu object={item}/>;
        }}
      />
    );
  }
}
