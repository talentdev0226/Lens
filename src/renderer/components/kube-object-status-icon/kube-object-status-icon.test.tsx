/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { getDiForUnitTesting } from "../../getDiForUnitTesting";
import type { DiRender } from "../test-utils/renderFor";
import { renderFor } from "../test-utils/renderFor";
import type { KubeObjectStatus } from "../../../common/k8s-api/kube-object-status";
import { KubeObjectStatusLevel } from "../../../common/k8s-api/kube-object-status";
import { KubeObject } from "../../../common/k8s-api/kube-object";
import { KubeObjectStatusIcon } from "./kube-object-status-icon";
import React from "react";
import { useFakeTime } from "../../../common/test-utils/use-fake-time";
import { getInjectable } from "@ogre-tools/injectable";
import type { DiContainer } from "@ogre-tools/injectable";
import { kubeObjectStatusTextInjectionToken } from "./kube-object-status-text-injection-token";
import { computed } from "mobx";

describe("kube-object-status-icon", () => {
  let render: DiRender;
  let di: DiContainer;

  beforeEach(() => {
    useFakeTime("2015-10-21T07:28:00Z");

    di = getDiForUnitTesting({ doGeneralOverrides: true });

    render = renderFor(di);
  });

  it("given no statuses, when rendered, renders as empty", () => {
    const kubeObject = getKubeObjectStub("irrelevant", "irrelevant");

    const { container } = render(<KubeObjectStatusIcon object={kubeObject} />);

    expect(container).toMatchSnapshot();
  });

  it('given level "critical" status, when rendered, renders with status', () => {
    const kubeObject = getKubeObjectStub("some-kind", "some-api-version");

    const statusTextInjectable = getStatusTextInjectable(
      KubeObjectStatusLevel.CRITICAL,
      "critical",
      "some-kind",
      ["some-api-version"],
    );

    di.register(statusTextInjectable);

    const { baseElement } = render(
      <KubeObjectStatusIcon object={kubeObject} />,
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('given level "info" status, when rendered, renders with status', () => {
    const kubeObject = getKubeObjectStub("some-kind", "some-api-version");

    const statusTextInjectable = getStatusTextInjectable(
      KubeObjectStatusLevel.INFO,
      "info",
      "some-kind",
      ["some-api-version"],
    );

    di.register(statusTextInjectable);

    const { baseElement } = render(
      <KubeObjectStatusIcon object={kubeObject} />,
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('given level "warning" status, when rendered, renders with status', () => {
    const kubeObject = getKubeObjectStub("some-kind", "some-api-version");

    const statusTextInjectable = getStatusTextInjectable(
      KubeObjectStatusLevel.WARNING,
      "warning",
      "some-kind",
      ["some-api-version"],
    );

    di.register(statusTextInjectable);

    const { baseElement } = render(
      <KubeObjectStatusIcon object={kubeObject} />,
    );

    expect(baseElement).toMatchSnapshot();
  });

  it("given status for all levels is present, when rendered, renders with statuses", () => {
    const kubeObject = getKubeObjectStub("some-kind", "some-api-version");

    const criticalStatusTextInjectable = getStatusTextInjectable(
      KubeObjectStatusLevel.CRITICAL,
      "critical",
      "some-kind",
      ["some-api-version"],
    );

    const warningStatusTextInjectable = getStatusTextInjectable(
      KubeObjectStatusLevel.WARNING,
      "warning",
      "some-kind",
      ["some-api-version"],
    );

    const infoStatusTextInjectable = getStatusTextInjectable(
      KubeObjectStatusLevel.INFO,
      "info",
      "some-kind",
      ["some-api-version"],
    );

    di.register(
      criticalStatusTextInjectable,
      warningStatusTextInjectable,
      infoStatusTextInjectable,
    );

    const { baseElement } = render(
      <KubeObjectStatusIcon object={kubeObject} />,
    );

    expect(baseElement).toMatchSnapshot();
  });

  it("given info and warning statuses are present, when rendered, renders with statuses", () => {
    const kubeObject = getKubeObjectStub("some-kind", "some-api-version");

    const warningStatusTextInjectable = getStatusTextInjectable(
      KubeObjectStatusLevel.WARNING,
      "warning",
      "some-kind",
      ["some-api-version"],
    );

    const infoStatusTextInjectable = getStatusTextInjectable(
      KubeObjectStatusLevel.INFO,
      "info",
      "some-kind",
      ["some-api-version"],
    );

    di.register(warningStatusTextInjectable, infoStatusTextInjectable);

    const { baseElement } = render(
      <KubeObjectStatusIcon object={kubeObject} />,
    );

    expect(baseElement).toMatchSnapshot();
  });


  it("given registration for wrong api version, when rendered, renders as empty", () => {
    const kubeObject = getKubeObjectStub("some-kind", "some-api-version");

    const statusTextInjectable = getStatusTextInjectable(
      KubeObjectStatusLevel.CRITICAL,
      "irrelevant",
      "some-kind",
      ["some-other-api-version"],
    );

    di.register(statusTextInjectable);

    const { baseElement } = render(
      <KubeObjectStatusIcon object={kubeObject} />,
    );

    expect(baseElement).toMatchSnapshot();
  });

  it("given registration for wrong kind, when rendered, renders as empty", () => {
    const kubeObject = getKubeObjectStub("some-kind", "some-api-version");

    const statusTextInjectable = getStatusTextInjectable(
      KubeObjectStatusLevel.CRITICAL,
      "irrelevant",
      "some-other-kind",
      ["some-api-version"],
    );

    di.register(statusTextInjectable);

    const { baseElement } = render(
      <KubeObjectStatusIcon object={kubeObject} />,
    );

    expect(baseElement).toMatchSnapshot();
  });

  it("given registration without status for exact kube object, when rendered, renders as empty", () => {
    const kubeObject = getKubeObjectStub("some-kind", "some-api-version");

    const statusTextInjectable = getInjectable({
      id: "some-id",
      instantiate: () => ({
        apiVersions: ["some-api-version"],
        kind: "some-kind",
        resolve: () => { return undefined as unknown as KubeObjectStatus; },
        enabled: computed(() => true),
      }),

      injectionToken: kubeObjectStatusTextInjectionToken,
    });

    di.register(statusTextInjectable);

    const { baseElement } = render(
      <KubeObjectStatusIcon object={kubeObject} />,
    );

    expect(baseElement).toMatchSnapshot();
  });
});

const getKubeObjectStub = (kind: string, apiVersion: string) => KubeObject.create({
  apiVersion,
  kind,
  metadata: {
    uid: "some-uid",
    name: "some-name",
    resourceVersion: "some-resource-version",
    namespace: "some-namespace",
    selfLink: "/foo",
  },
});

const getStatusTextInjectable = (level: KubeObjectStatusLevel, title: string, kind: string, apiVersions: string[]) => getInjectable({
  id: title,
  instantiate: () => ({
    apiVersions,
    kind,

    resolve: (kubeObject: KubeObject) => ({
      level,
      text: `Some ${title} status for ${kubeObject.getName()}`,
      timestamp: "2015-10-19T07:28:00Z",
    }),

    enabled: computed(() => true),
  }),

  injectionToken: kubeObjectStatusTextInjectionToken,
});
