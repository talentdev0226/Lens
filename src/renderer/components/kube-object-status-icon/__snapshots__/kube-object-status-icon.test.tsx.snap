// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kube-object-status-icon given info and warning statuses are present, when rendered, renders with statuses 1`] = `
<body>
  <div>
    <i
      class="Icon KubeObjectStatusIcon warning material focusable"
      id="tooltip_target_5"
    >
      <span
        class="icon"
        data-icon-name="warning"
      >
        warning
      </span>
      <div />
    </i>
  </div>
  <div
    class="Tooltip narrow formatter"
  >
    <div
      class="KubeObjectStatusTooltip"
    >
      <div
        class="level warning"
      >
        <span
          class="title"
        >
          Warning
        </span>
        <div
          class="status msg"
        >
          - 
          Some warning status for some-name
           
          <span
            class="age"
          >
             · 
            2d
          </span>
        </div>
      </div>
      <div
        class="level info"
      >
        <span
          class="title"
        >
          Info
        </span>
        <div
          class="status msg"
        >
          - 
          Some info status for some-name
           
          <span
            class="age"
          >
             · 
            2d
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`kube-object-status-icon given level "critical" status, when rendered, renders with status 1`] = `
<body>
  <div>
    <i
      class="Icon KubeObjectStatusIcon error material focusable"
      id="tooltip_target_1"
    >
      <span
        class="icon"
        data-icon-name="error"
      >
        error
      </span>
      <div />
    </i>
  </div>
  <div
    class="Tooltip narrow formatter"
  >
    <div
      class="KubeObjectStatusTooltip"
    >
      <div
        class="level error"
      >
        <span
          class="title"
        >
          Critical
        </span>
        <div
          class="status msg"
        >
          - 
          Some critical status for some-name
           
          <span
            class="age"
          >
             · 
            2d
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`kube-object-status-icon given level "info" status, when rendered, renders with status 1`] = `
<body>
  <div>
    <i
      class="Icon KubeObjectStatusIcon info material focusable"
      id="tooltip_target_2"
    >
      <span
        class="icon"
        data-icon-name="info"
      >
        info
      </span>
      <div />
    </i>
  </div>
  <div
    class="Tooltip narrow formatter"
  >
    <div
      class="KubeObjectStatusTooltip"
    >
      <div
        class="level info"
      >
        <span
          class="title"
        >
          Info
        </span>
        <div
          class="status msg"
        >
          - 
          Some info status for some-name
           
          <span
            class="age"
          >
             · 
            2d
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`kube-object-status-icon given level "warning" status, when rendered, renders with status 1`] = `
<body>
  <div>
    <i
      class="Icon KubeObjectStatusIcon warning material focusable"
      id="tooltip_target_3"
    >
      <span
        class="icon"
        data-icon-name="warning"
      >
        warning
      </span>
      <div />
    </i>
  </div>
  <div
    class="Tooltip narrow formatter"
  >
    <div
      class="KubeObjectStatusTooltip"
    >
      <div
        class="level warning"
      >
        <span
          class="title"
        >
          Warning
        </span>
        <div
          class="status msg"
        >
          - 
          Some warning status for some-name
           
          <span
            class="age"
          >
             · 
            2d
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`kube-object-status-icon given no statuses, when rendered, renders as empty 1`] = `<div />`;

exports[`kube-object-status-icon given registration for wrong api version, when rendered, renders as empty 1`] = `
<body>
  <div />
</body>
`;

exports[`kube-object-status-icon given registration for wrong kind, when rendered, renders as empty 1`] = `
<body>
  <div />
</body>
`;

exports[`kube-object-status-icon given registration without status for exact kube object, when rendered, renders as empty 1`] = `
<body>
  <div />
</body>
`;

exports[`kube-object-status-icon given status for all levels is present, when rendered, renders with statuses 1`] = `
<body>
  <div>
    <i
      class="Icon KubeObjectStatusIcon error material focusable"
      id="tooltip_target_4"
    >
      <span
        class="icon"
        data-icon-name="error"
      >
        error
      </span>
      <div />
    </i>
  </div>
  <div
    class="Tooltip narrow formatter"
  >
    <div
      class="KubeObjectStatusTooltip"
    >
      <div
        class="level error"
      >
        <span
          class="title"
        >
          Critical
        </span>
        <div
          class="status msg"
        >
          - 
          Some critical status for some-name
           
          <span
            class="age"
          >
             · 
            2d
          </span>
        </div>
      </div>
      <div
        class="level warning"
      >
        <span
          class="title"
        >
          Warning
        </span>
        <div
          class="status msg"
        >
          - 
          Some warning status for some-name
           
          <span
            class="age"
          >
             · 
            2d
          </span>
        </div>
      </div>
      <div
        class="level info"
      >
        <span
          class="title"
        >
          Info
        </span>
        <div
          class="status msg"
        >
          - 
          Some info status for some-name
           
          <span
            class="age"
          >
             · 
            2d
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;
